.builders/upload.py

```diff
@@ -0,0 +1,164 @@
+def is_valid_project_name(project_name: str) -> bool:
+    # https://peps.python.org/pep-0508/#names
+    return VALID_PROJECT_NAME.search(project_name) is not None
+
+
+def normalize_project_name(name: str) -> str:
+    # https://peps.python.org/pep-0503/#normalized-names
+    return UNNORMALIZED_PROJECT_NAME_CHARS.sub('-', name).lower()
+
+
+def extract_metadata(wheel: Path) -> email.Message:
+    with ZipFile(str(wheel)) as zip_archive:
+        for path in zip_archive.namelist():
+            root = path.split('/', 1)[0]
+            if root.endswith('.dist-info'):
+                dist_info_dir = root
+                break
+        else:
+            message = f'Could not find the `.dist-info` directory in wheel: {wheel.name}'
+            raise RuntimeError(message)
+
+        try:
+            with zip_archive.open(f'{dist_info_dir}/METADATA') as zip_file:
+                metadata_file_contents = zip_file.read().decode('utf-8')
+        except KeyError:
+            message = f'Could not find a `METADATA` file in the `{dist_info_dir}` directory'
+            raise RuntimeError(message) from None
+
+    return email.message_from_string(metadata_file_contents)
+
+
+def display_message_block(message: str) -> None:
+    divider = f'+{"-" * (len(message) + 2)}+'
+    print(divider)
+    print(f'| {message} |')
+    print(divider)
+
+
+def main():
+    parser = argparse.ArgumentParser(prog='builder', allow_abbrev=False)
+    parser.add_argument('wheel_dir')
+    args = parser.parse_args()
+
+    client = storage.Client()
+    bucket = client.bucket(BUCKET_NAME)
+    artifact_types: list[str] = []
+    for entry in sorted(Path(args.wheel_dir).iterdir(), key=lambda p: p.name):
+        artifact_type = entry.name
+        artifact_types.append(artifact_type)
+        message = f'Processing {artifact_type} wheels'
+        display_message_block(message)
+
+        upload_data: tuple[str, email.Message, Path] = []
+        for wheel in entry.iterdir():
+            project_metadata = extract_metadata(wheel)
+            project_name = project_metadata['Name']
+            if not is_valid_project_name(project_name):
+                message = f'Invalid project name `{project_name}` found in wheel: {wheel.name}'
+                raise RuntimeError(message)
+
+            upload_data.append((normalize_project_name(project_name), project_metadata, wheel))
+
+        queued = len(upload_data)
+        upload_data.sort()
+
+        for i, (project_name, project_metadata, wheel) in enumerate(upload_data, start=1):
+            prefix = f'({i}/{queued})'
+            padding = ' ' * (len(prefix) + 1)
+
+            artifact = bucket.blob(f'{artifact_type}/{project_name}/{wheel.name}')
+
+            # PyPI artifacts never change, so we don't need to upload them again.
+            if artifact_type == 'external' and artifact.exists():
+                print(f'{prefix} {project_name}=={project_metadata["Version"]} already exists')
+                continue
```

Comment:
Not necessarily for this PR, but we should probably consider the scenario where a wheel changes without updating the package version due to a change in the build images / build dependencies.

Reply:
Actually this part is only for wheels that we do not build and wheels we do build will always overwrite. I did that explicitly to account for the edge case that you mention.

Reply:
True, I see that now. I'm not sure overwriting a wheel is a good policy though because it might break the Agent build as the hash will change in the repo before the lockfile is actually updated, no?

Reply:
This should be fine since changes to `.builders/` trigger the workflow.